File: src/pocketcloud/cloud/provider/CloudProvider.php
<?php

namespace pocketcloud\cloud\provider;

use pocketcloud\cloud\config\impl\MainConfig;
use pocketcloud\cloud\group\ServerGroup;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\util\promise\Promise;

abstract class CloudProvider {

    private static ?self $current = null;

    abstract public function addTemplate(Template $template): void;

    abstract public function removeTemplate(Template $template): void;

    abstract public function editTemplate(Template $template, array $newData): void;

    abstract public function getTemplate(string $template): Promise;

    abstract public function checkTemplate(string $template): Promise;

    abstract public function getTemplates(): Promise;

    abstract public function addServerGroup(ServerGroup $serverGroup): void;

    abstract public function removeServerGroup(ServerGroup $serverGroup): void;

    abstract public function editServerGroup(ServerGroup $serverGroup, array $newData): void;

    abstract public function getServerGroup(string $serverGroup): Promise;

    abstract public function checkServerGroup(string $serverGroup): Promise;

    abstract public function getServerGroups(): Promise;

    abstract public function setModuleState(string $module, bool $enabled): void;

    abstract public function getModuleState(string $module): Promise;

    abstract public function enablePlayerNotifications(string $player): void;

    abstract public function disablePlayerNotifications(string $player): void;

    abstract public function hasNotificationsEnabled(string $player): Promise;

    abstract public function addToWhitelist(string $player): void;

    abstract public function removeFromWhitelist(string $player): void;

    abstract public function isOnWhitelist(string $player): Promise;

    abstract public function getWhitelist(): Promise;

    public static function select(): void {
        self::$current = match (MainConfig::getInstance()->getProvider()) {
            "mysql" => new CloudMySqlProvider(),
            default => new CloudJsonProvider()
        };
    }

    public static function current(): self {
        if (self::$current === null) self::select();
        return self::$current;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/provider/CloudProvider.java
package de.pocketcloud.cloud.provider;

import de.pocketcloud.api.component.group.IServerGroup;
import de.pocketcloud.api.component.template.ITemplate;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.common.concurrent.Promise;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public abstract class CloudProvider {

    private static CloudProvider current = null;

    public abstract Promise<Void> addTemplate(ITemplate template);

    public abstract Promise<Void> removeTemplate(ITemplate template);

    public abstract Promise<Void> editTemplate(ITemplate template, Map<String, Object> newData);

    public abstract Promise<Optional<Template>> getTemplate(String template);

    public abstract Promise<Boolean> checkTemplate(String template);

    public abstract Promise<Map<String, Template>> getTemplates();

    public abstract Promise<Void> addServerGroup(IServerGroup serverGroup);

    public abstract Promise<Void> removeServerGroup(IServerGroup serverGroup);

    public abstract Promise<Void> editServerGroup(IServerGroup serverGroup, Map<String, Object> newData);

    public abstract Promise<Optional<ServerGroup>> getServerGroup(String serverGroup);

    public abstract Promise<Boolean> checkServerGroup(String serverGroup);

    public abstract Promise<Map<String, ServerGroup>> getServerGroups();

    public abstract Promise<Void> enablePlayerNotifications(String player);

    public abstract Promise<Void> disablePlayerNotifications(String player);

    public abstract Promise<Boolean> hasNotificationsEnabled(String player);

    public abstract Promise<List<String>> getNotificationList();

    public abstract Promise<Void> addToWhitelist(String player);

    public abstract Promise<Void> removeFromWhitelist(String player);

    public abstract Promise<Boolean> isOnWhitelist(String player);

    public abstract Promise<List<String>> getWhitelist();

    public static void select() {
        String provider = PocketCloud.instance().config().provider();
        current = switch (provider) {
            case "mysql" -> new CloudMySqlProvider();
            default -> new CloudJsonProvider();
        };
    }

    public static CloudProvider current() {
        if (current == null) select();
        return current;
    }
}